src/main/java/com/google/devtools/build/lib/analysis/RepoMappingManifestAction.java

```diff
@@ -0,0 +1,122 @@
+import net.starlark.java.eval.EvalException;
+
+/** Creates a manifest file describing the repos and mappings relevant for a runfile tree. */
+public class RepoMappingManifestAction extends AbstractFileWriteAction {
+  private static final UUID MY_UUID = UUID.fromString("458e351c-4d30-433d-b927-da6cddd4737f");
+
+  private final ImmutableSortedMap<RepositoryName, ImmutableSortedMap<String, RepositoryName>>
+      reposAndMappings;
+  private final String workspaceName;
+
+  public RepoMappingManifestAction(
+      ActionOwner owner,
+      Artifact output,
+      Map<RepositoryName, RepositoryMapping> reposAndMappings,
+      String workspaceName) {
+    super(owner, NestedSetBuilder.emptySet(Order.STABLE_ORDER), output, /*makeExecutable=*/ false);
+    this.reposAndMappings =
+        ImmutableSortedMap.copyOf(
+            Maps.transformValues(
+                reposAndMappings, repoMapping -> ImmutableSortedMap.copyOf(repoMapping.entries())),
+            (a, b) -> OsPathPolicy.getFilePathOs().compare(a.getName(), b.getName()));
+    this.workspaceName = workspaceName;
+  }
+
+  @Override
+  public String getMnemonic() {
+    return "RepoMappingManifest";
+  }
+
+  @Override
+  protected String getRawProgressMessage() {
+    return "writing repo mapping manifest for " + getOwner().getLabel();
+  }
+
+  @Override
+  protected void computeKey(
+      ActionKeyContext actionKeyContext,
+      @Nullable ArtifactExpander artifactExpander,
+      Fingerprint fp)
+      throws CommandLineExpansionException, EvalException, InterruptedException {
+    fp.addUUID(MY_UUID);
+    fp.addString(workspaceName);
+    for (Entry<RepositoryName, ImmutableSortedMap<String, RepositoryName>> repoAndMapping :
+        reposAndMappings.entrySet()) {
+      fp.addString(repoAndMapping.getKey().getName());
+      fp.addInt(repoAndMapping.getValue().size());
+      for (Entry<String, RepositoryName> mappingEntry : repoAndMapping.getValue().entrySet()) {
+        fp.addString(mappingEntry.getKey());
+        fp.addString(mappingEntry.getValue().getName());
+      }
+    }
+  }
+
+  @Override
+  public DeterministicWriter newDeterministicWriter(ActionExecutionContext ctx)
+      throws InterruptedException, ExecException {
+    return out -> {
+      Writer writer = new BufferedWriter(new OutputStreamWriter(out, ISO_8859_1));
+      for (Entry<RepositoryName, ImmutableSortedMap<String, RepositoryName>> repoAndMapping :
+          reposAndMappings.entrySet()) {
+        for (Entry<String, RepositoryName> mappingEntry : repoAndMapping.getValue().entrySet()) {
+          if (mappingEntry.getKey().isEmpty()) {
+            // The apparent repo name can only be empty for the main repo. We skip this line.
+            continue;
+          }
+          writer.write(repoAndMapping.getKey().getName());
+          writer.write(',');
+          writer.write(mappingEntry.getKey());
+          writer.write(',');
+          if (mappingEntry.getValue().isMain()) {
```

Comment:
nit: a comment might be useful here that explains why this is needed. Let me see if my understanding is correct: the name of the main repository is `""`, but runfiles in it appear in the runfiles tree under the directory whose name is the same as the name of the main workspace. Is that so?

Reply:
done. (yes you're correct)